ddev/ddev
# refactor: Refactor `ddev get` to `ddev add-on` and into sub-commands, fixes #6146

## The Issue

- #6146



`ddev get` isn't immediately obvious by the name what it does, and it seems to be doing three separate things (installing, removing, and listing add-ons).

## How This PR Solves The Issue

Creates an umbrella command `ddev add-on`, which has a sub-command for each of the three responsibilities of the `ddev get` command:
1. `ddev add-on get`
2. `ddev add-on remove`
3. `ddev add-on list`

Appropriate aliases are added - e.g. `ddev add-on`, `ddev addon`, `ddev add-ons`, and `ddev addons` all work. This means people don't have to remember whether it's plural or not, nor whether it's hyphenated (I forgot both of those while I was testing this PR, so this makes it easy to use).

The existing `ddev get` command is turned into an alias for the various `ddev add-on` commands, so the code only exists in one place. It is also marked as deprecated, so that users get a message to use the new `ddev add-on` command instead.

In a future release, once people have gotten used to using `ddev add-on`, the `ddev get` command can be removed.

## Manual Testing Instructions

Rendered docs at https://ddev--6482.org.readthedocs.build/en/6482/users/extend/additional-services/ (and scattered around, look carefully)

Try `ddev add-on` on its own - it should just list usage info.

Try `ddev add-on` with flags and arguments. It should still just list usage info.

Try `ddev add-on get <add-on>` to install some add-ons.

Try `ddev add-on remove <add-on>` to remove some installed add-ons.

Try `ddev add-on list` with and without the various flags to list official, unnoficial, and installed add-ons.

Try the old `ddev get` command - that should still work exactly as expected, just with the additional output at the start saying to use the new command instead.

Check the usage info for each of the `ddev add-on` sub-commands is correct.

## Automated Testing Overview



I've duplicated the existing tests for `ddev get`. This means we are still testing `ddev get` to ensure that the alias works as expected, and we're also testing the new `ddev add-on` sub-commands.

This does mean the tests are in two places at once, so a change to the tests in one place needs to also be made in the other place. I couldn't find a clean way to alias the tests.

If you think it's better to just test `ddev add-on` with the existing test coverage, and instead just write some minimal tests for the `ddev get` alias, I'm happy to make that change.

## Release/Deployment Notes



Everything that used to work still works. Nothing will break as a result of this PR.

We should highlight that `ddev add-on` is the new best practice, and that `ddev get` will eventually be removed.

People should update any scripts they have that use `ddev get` to instead use the relevant `ddev add-on` sub-command.

People should update the instructions in their add-on readme files to use `ddev add-on get` instead of `ddev get`.
